### DISCUSSION TITLE
Combining multiple background blend modes

### DISCUSSION THREAD

**Discussion by ricardobalk** [tailwindlabs/tailwindcss#11581] [Help] [answered]
I'm currently exploring different solutions for combining multiple background blend modes in TailwindCSS. I have already attempted to use different `@apply` directives with no success. Another approach I have considered is creating a custom `@apply` directive using `tailwind.config.js`, such as `@apply bg-blend-custom`, but I'm not satisfied with that as a solution.

In CSS, you can combine background blend modes, so each background image or gradient can have its own blend mode applied, resulting in unique and dynamic visual effects.

```css
background: url('first.jpg'), url('second.jpg');
background-blend-mode: multiply, screen;
```

But I want this with TailwindCSS, so I can also combine it with `dark:bg-...`, which is based on a class name, as defined in my `tailwind.config.js`.

Things I have tried, without achieving the desired result:

 - `@apply dark:bg-blend-[multiply,normal];`
 - `@apply dark:bg-blend[multiply,normal];`
 - `@apply dark:bg-[blend:multiply,normal];`
 - Exploring the option of creating a custom directive to handle the multiple background blend modes, such as @apply bg-blend-custom, is technically possible, but it is personally not my most preferred solution.

Expected Behavior: I expect to find a solution within TailwindCSS that allows me to conveniently combine multiple background blend modes.

Please provide any insights, guidance, or alternative solutions on how to achieve this effect with TailwindCSS.

Thank you for your assistance!

**Comment by wongjn** [comment]
Since it seems like you are writing in a stylesheet, you should consider using vanilla CSS:

```css
background: url('first.jpg'), url('second.jpg');
background-blend-mode: multiply, screen;
```

Otherwise, consider applying the properties directly as classes on the elements themselves, like:

```html
<div class="bg-[url(first.jpg),url(second.jpg)] [background-blend-mode:multiply,screen]">
```
As you say, you can abstract this to configuration:

```js
module.exports = {
  theme: {
    extend: {
      backgroundImage: {
        foo: 'url(first.jpg), url(second.jpg)',
      },
      backgroundBlend: {
        foo: 'multiply,screen',
      },
    },
  },
  plugins: [
    require('tailwindcss/plugin')(({ matchUtilities, theme }) => {
      matchUtilities(
        {
          'bg-blend': (value) => ({
            backgroundBlendMode: value,
          }),
        },
        {
          values: theme('backgroundBlend'),
        }
      );
    }),
  ],
};
```
```html
<div class="bg-foo bg-blend-foo">
```

**Reply by ricardobalk** [reply]
This is like exactly the solution I try to avoid. I am using TailwindCSS with PostCSS in a Vue project, like so:
```javascript
/* tailwind.config.js */
const defaultTheme = require('tailwindcss/defaultTheme')
const plugin = require('tailwindcss/plugin');
/** @type {import('tailwindcss').Config} */
module.exports = {
  content: [
    "./components/**/*.{js,vue,ts}",
    "./layouts/**/*.vue",
    "./pages/**/*.vue",
    "./plugins/**/*.{js,ts}",
    "./nuxt.config.{js,ts}",
    "./app.vue",
  ],
  darkMode: 'class',
  plugins: [
    plugin(function({ addUtilities }) {
      addUtilities({
        '.bg-blend-custom-1': {
          'background-blend-mode': 'multiply, normal',
        },
        '.bg-blend-custom-2': {
          'background-blend-mode': 'soft-light, normal',
        },
      })
    })
  ],
}
```

```vue
<style lang="postcss">
.class {
  @apply w-full text-white bg-no-repeat bg-[#6bffdc] bg-[contain,cover];

  @apply dark:mb-12 dark:bg-slate-900 dark:text-slate-300;
  @apply bg-blend-custom-1 dark:bg-blend-custom-2 dark:py-8 dark:text-slate-400;
}
</style>
```

Important is that I'm using `darkMode: class`, otherwise I would just yeet my dark styles inside a media query and indeed &mdash; use custom css for it.

But also, using this `bg-blend-custom-1` seems a bit odd, as it is possible to combine other background properties like the one of `bg-[contain,cover]` &mdash; so I expected the same functionality with `bg-blend-*` 🤔

**Comment by ricardobalk** [comment, marked as answer]
Aha, found a way, partially by your answer, @wongjn, and although it's not the most clean solution, it's better than adding a custom directive to `tailwind.config.js`

```vue
<style lang="postcss">
.class {
  @apply [background-blend-mode:normal,_normal] dark:[background-blend-mode:soft-light,_normal];
}
</style>
```

Where the `_` will be converted to a space. (I have read that in the docs earlier this week).
